Don't try aliasing D keywords

mihails-strasuns-sociomantic opened the pull request:
Most common issue this prevents is trying to alias `assert` keyword when
translating C macro.

mihails-strasuns-sociomantic commented:
Green

jacob-carlborg commented on dstep/core/Traits.d:
```diff
@@ -0,0 +1,30 @@
+    ];
+ 
+    return DKeywords.canFind(ident);
+}
```
There is already an existing function for this: https://github.com/jacob-carlborg/dstep/blob/master/dstep/translator/Translator.d#L487

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -2053,6 +2053,10 @@ bool translateFunctAlias(
 
+        import dstep.core.Traits : isKeyword;
+        if (ident.spelling.isKeyword())
+            return false;
```
In other places in the code, when there's a symbol with is a D keyword that is translate, an underscore is added as a suffix. Which there is a separate function for [1].

[1] https://github.com/jacob-carlborg/dstep/blob/master/dstep/translator/Translator.d#L469

mihails-strasuns-sociomantic replied:
This case is a bit different - you simply can't alias a keyword. For example `alias foo = assert` is not legal D code, no matter what you do.

jacob-carlborg replied:
I guess that would end up as `alias foo = assert_`. But that might be a bit pointless.

mihails-strasuns-sociomantic replied:
Not only pointless, it won't even compile because no one defines `assert_`.

jacob-carlborg replied:
But with this change, it's not possible to handle when it _is_ defined. Take this example from test suite, slightly modified:

```c
#define delegate(major, minor)#major"."#minor
#define STRINGIZE(major, minor)delegate(major, minor)
```

Currently this is translated to this:

```d
extern (D) string delegate(T0, T1)(auto ref T0 major, auto ref T1 minor)
{
    import std.conv : to;

    return to!string(major) ~ "." ~ to!string(minor);
}

alias STRINGIZE = delegate;
```

Which is, of course, a bug. Instead it should be translated with a `_` suffix for identifier `delegate`:

```d
extern (D) string delegate_(T0, T1)(auto ref T0 major, auto ref T1 minor)
{
    import std.conv : to;

    return to!string(major) ~ "." ~ to!string(minor);
}

alias STRINGIZE = delegate_;
```

mihails-strasuns-sociomantic replied:
I see, there seems to be a bit of a clash between your example and the one I have. To be more, specific, issue I am trying to solve is translation of this macro:

```c
#define PROTOBUF_C_ASSERT(condition) assert(condition)
```

Here `assert` resolves to function in `assert.h` in C header, but exiting translation will try turning it into `alias PROTOBUF_C_ASSERT = asser` which of course can't work.

jacob-carlborg replied:
Looking at this problem in general, I don't see why adding a `_` suffix wouldn't work. Example:

```c
void delegate(bool condition);
#define PROTOBUF_C_DELEGATE(condition) delegate(condition)
```

I don't see why that couldn't be translated to:

```d
pragma(mangle, "delegate") void delegate_(bool condition);
alias PROTOBUF_C_DELEGATE = delegate_;
```

Your assumption in this PR is also that `assert` will always be the symbol defined in `assert.h`, which might not be the case. At least how the code is written now, it will ignore all symbols that are keywords in D, which might not be desirable.

In this particular case with `assert`, I don't think anyone would think it's useful to translate that function/macro to D (a corresponding `assert` module does not exist in druntime). If someone would manually translate some C code using `assert` they would most likely translate the use of `assert` to the D built-in `assert` and not translate the actual macro/function.

I'm not sure what's the best short term solution is. One idea would be to special case `assert` and implement the `assert` macro inline, or completely ignore it.

For a long term solution, I've been thinking for a while about a generic way to specify declarations that should be ignored.

jacob-carlborg commented:
There are not tests.

mathias-lang-sociomantic commented:
Ping @mihails-strasuns-sociomantic

mihails-strasuns-sociomantic commented:
Updated. After some more thinking about the issue I agree that simply special casing "assert" alone to generate function instead of an alias is most practical approach.

Also added test case.

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -2053,6 +2053,15 @@ bool translateFunctAlias(
     {
         Identifier ident = cast(Identifier) expr.expr;
 
+        if (ident.spelling == "assert")
```
Is it worth checking that it's defined in `assert.h` as well?

ciechowoj replied:
We would need to be able to determine if it's reachable 'through' the `assert.h` header...

jacob-carlborg replied:
> Does libclang expose such information? It is not well-defined in C in general because of copy-paste nature of headers

Yes, at least for regular non-macro symbols. It's possible to get the location/path of a cursor [1], not sure if macros are represented as cursors. It's used extensively by the include handler.

[1] https://github.com/jacob-carlborg/dstep/blob/master/clang/Cursor.d#L68

jacob-carlborg replied:
> We would need to be able to determine if it's reachable 'through' the assert.h header

@ciechowoj are you referring to these platform specific header files again? The true location of a symbol doesn't match the documentation/standard. I hate those.

ciechowoj replied:
Yes, that's what I'm referring to.

mathias-lang-sociomantic replied:
> Is it worth checking that it's defined in assert.h as well?

Given the discussion, I would say "no" :)

mihails-strasuns-sociomantic replied:
Yeah I checked how `assert.h` header looks on few systems and it looks like euristic that is doomed to fail. Unless libclang provides some super-magical helpers to do that I think assuming that any `assert` no matter where it comes from acts like D built-in `assert` is reasonable.

jacob-carlborg replied:
Hmm, I see.

mihails-strasuns-sociomantic commented:
Is there anything else that needs to be done with this PR?

jacob-carlborg commented on dstep/translator/MacroDefinition.d:
```diff
@@ -2053,6 +2053,15 @@ bool translateFunctAlias(
+        {
+            // aliasing `assert` from C "assert.h" is commong enough to warrant
+            // a special case to use function instead of an alias
+            output.singleLine("void %s(T...)(T args) { assert(args); }",
```
* It's only possible to pass two arguments two `assert`, a boolean condition and a string
* Please don't use this compact syntax for the generated function

mihails-strasuns-sociomantic commented:
Updated:
- `if` constraint for amount of `assert` arguments
- multi-line output for the function

jacob-carlborg commented:
👌
